File: LeetCode/Explore/TopInterview/Blind75/DynamicProgramming/91_decode_ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        dp = [0] * (len(s) + 1)
        dp[0] = 1
        if s[0] == '0':
          dp[1] = 0
        else:
          dp[1] = 1
          
        for i in range(2, len(s) + 1, 1):
          if int(s[i - 1:i]) >= 1:
            dp[i] = dp[i - 1]
          if int(s[i -2:i]) >= 10 and int(s[i - 2:i]) <= 26:
            dp[i] += dp[i - 2]
            
        return dp[len(s)]

File: LeetCode/Explore/TopInterview/Meta/DynamicProgramming/91_decode_ways.cpp
class Solution {
public:
    int numDecodings(string s) {
      vector<int> dp(s.size() + 1, 0);
      dp[0] = 1;
      dp[1] = (s[0] == '0') ? 0 : 1;
      for (int i = 2; i <= s.size(); i++) {
        if (stoi(s.substr(i - 1, 1)) >= 1) {
          dp[i] = dp[i - 1];
        }
        if (stoi(s.substr(i - 2, 2)) >= 10 && 
           stoi(s.substr(i - 2, 2)) <= 26) {
          dp[i] += dp[i - 2];
        }
      }
      return dp[s.size()];
    }
};
